File: DotNetTreeShadows/Models/GameModel/PlayerBoard.cs
using System;
using System.Linq;
using dotnet_tree_shadows.Models.Enums;
using static dotnet_tree_shadows.Models.GameModel.PlayerBoard.PieceCount;

namespace dotnet_tree_shadows.Models.GameModel {
  public class PlayerBoard {

    private PieceCount[] PieceCounts { get; } = {
      //pieceType, available, onboard, max
      //value,offset, mask
      
      new PieceCount( PieceType.Seed, new BitValue( 2, 8, 15 ), new BitValue( 4, 12, 7 ), 4 ),
      new PieceCount( PieceType.SmallTree, new BitValue( 4, 15, 15 ), new BitValue( 4, 19, 7 ), 4 ),
      new PieceCount( PieceType.MediumTree, new BitValue( 1, 22, 7 ), new BitValue( 3, 25, 7 ), 3 ),
      new PieceCount( PieceType.LargeTree, new BitValue( 0, 28, 3 ), new BitValue( 2, 30, 3 ), 2 )
    };
    
    

    public PlayerBoard () { }

    public PlayerBoard (int boardCode) {
      Light = (int) ((boardCode >> 2) & 0b_11_1111);
      TreeType = (TreeType) (boardCode & 0b_11);

      foreach ( PieceCount pieceCount in PieceCounts ) {
        pieceCount.ParseBoardCode( boardCode );
      }
    }


    public int BoardCode {
      get {
        int boardCode = PieceCounts.Aggregate(
            treeTypeCode | lightCode,
            (current, pieceCount) => current | pieceCount.pieceCode
          );

        return boardCode;
      }
    }

    public TreeType TreeType { get; set; } = TreeType.Ash;

    private int treeTypeCode {
      get => (int) TreeType;
    }

    private int light = 0;
    
    public int Light {
      get => light;
      set =>
        light = value > 30
                  ? 30
                  : value;
    }

    private int lightCode {
      get => (int) Light << 2;
    }

    public PieceCount Pieces (PieceType pieceType) => PieceCounts[(int) pieceType];

    public void SpendLight (int i) => Light -= i;
    public void RecoverLight (int i) => Light += i;

    public class PieceCount {

      private BitValue available;
      private BitValue onPlayerBoard;
      private readonly int maxOnBoard;
      private readonly int[] prices;

      public PieceCount (PieceType pieceType, BitValue available, BitValue onPlayerBoard, int maxOnBoard) {
        this.available = available;
        this.onPlayerBoard = onPlayerBoard;
        this.maxOnBoard = maxOnBoard;
        prices = pieceType switch {
          PieceType.Seed => new[] { 2, 2, 1, 1 },
          PieceType.SmallTree => new[] { 3, 3, 2, 2 },
          PieceType.MediumTree => new[] { 4, 3, 3 },
          PieceType.LargeTree => new[] { 5, 4 },
          _ => throw new ArgumentException( "invalid pieceType" )
        };
      }

      public int pieceCode {
        get => (Available << available.Offset) | (OnPlayerBoard << onPlayerBoard.Offset);
      }

      public int Available {
        get => available.Value;
        private set => available.Value = value;
      }

      public int OnPlayerBoard {
        get => onPlayerBoard.Value;
        private set => onPlayerBoard.Value = Math.Min(value, maxOnBoard);
      }

      public int NextPrice {
        get =>
          OnPlayerBoard == 0
            ? int.MaxValue
            : prices[OnPlayerBoard - 1];
      }

      public void IncreaseAvailable () => Available++;
      public void DecreaseAvailable () => Available--;
      public void IncreaseOnPlayerBoard () => OnPlayerBoard++;
      public void DecreaseOnPlayerBoard () => OnPlayerBoard--;
      public bool CanReturnSafely () => OnPlayerBoard < maxOnBoard;

      public void ParseBoardCode (int boardCode) {
        Available = (boardCode >> available.Offset) & available.Mask;
        OnPlayerBoard = (boardCode >> onPlayerBoard.Offset) & onPlayerBoard.Mask;
      }

      public struct BitValue {

        public BitValue (int value, int offset, int mask) {
          Value = value;
          Offset = offset;
          Mask = mask;
        }

        public int Value { get; set; }
        public int Offset { get; set; }
        public int Mask { get; set; }

      }

    }

    public static PlayerBoard Get (Game game, string playerId) => new PlayerBoard(game.PlayerBoards[playerId]);

    public static Game Set (Game game, string playerId, PlayerBoard playerBoard) {
      game.PlayerBoards[playerId] = playerBoard.BoardCode;
      return game;
    }

  }
}

File: DotNetTreeShadows/client/src/store/game/types/playerBoard.ts
import {TreeType} from "../../board/types/treeType";
import {PieceType} from "../../board/types/pieceType";

interface PieceCount {
    available: number;
    onPlayerBoard: number;
    max: number;
    increaseAvailable: () => number;
    decreaseAvailable: () => number;
    increaseOnPlayerBoard: () => number;
    decreaseOnPlayerBoard: () => number;
}


export default class PlayerBoard {
    public static GetLight = (boardCode: number): number => {
        return (boardCode >> 2) & 63;
    }

    public static TreeType = (boardCode: number): TreeType => boardCode & 3 as TreeType;

    public static seeds = (boardCode: number): PieceCount => PlayerBoard.DecodePieceCounts(boardCode, [8, 15], [12, 7], 4)
    public static smallTrees = (boardCode: number): PieceCount => PlayerBoard.DecodePieceCounts(boardCode, [15, 15], [19, 7], 4)
    public static mediumTrees = (boardCode: number): PieceCount => PlayerBoard.DecodePieceCounts(boardCode, [22, 7], [25, 7], 3)
    public static largeTrees = (boardCode: number): PieceCount => PlayerBoard.DecodePieceCounts(boardCode, [28, 3], [30, 3], 2)

    public static GetPieces = (boardCode: number, pieceType: PieceType): PieceCount => {
        switch (pieceType) {
            case PieceType.Seed:
                return PlayerBoard.seeds(boardCode);
            case PieceType.SmallTree:
                return PlayerBoard.smallTrees(boardCode);
            case PieceType.MediumTree:
                return PlayerBoard.mediumTrees(boardCode);
            case PieceType.LargeTree:
                return PlayerBoard.largeTrees(boardCode);
        }
        throw new Error(`Invalid pieceType: ${pieceType} - ${PieceType[pieceType]} `)
    }

    public static spaces = (pieceType: PieceType) => {
        switch (pieceType) {
            case PieceType.Seed:
                return 4;
            case PieceType.SmallTree:
                return 4;
            case PieceType.MediumTree:
                return 3;
            case PieceType.LargeTree:
                return 2;
        }
    }

    public static prices = (pieceType: PieceType): number[] => {
        switch (pieceType) {
            case PieceType.Seed:
                return [2, 2, 1, 1];
            case PieceType.SmallTree:
                return [3, 3, 2, 2];
            case PieceType.MediumTree:
                return [4, 3, 3];
            case PieceType.LargeTree:
                return [5, 4];
        }
    };

    public static MakeGrid = (boardCode: number) => {
        const pieceTypes: PieceType[] = [PieceType.Seed, PieceType.SmallTree, PieceType.MediumTree, PieceType.LargeTree];
        const grid: PieceDetails[][] = [];
        pieceTypes.forEach(pt => {
            const column: PieceDetails[] = [];
            const spaces = PlayerBoard.spaces(pt);
            const onPlayerBoard = PlayerBoard.GetPieces(boardCode, pt).onPlayerBoard;
            for (let i = 0; i < spaces; i++) {
                const status = i <= onPlayerBoard - 1
                        ? "Filled"
                        : "Empty";
                column.push({
                    status: status,
                    price: PlayerBoard.prices(pt)[i].toString() ?? "",
                    key: PieceType[pt]+i
                });
            }

            grid.push(column);

        });
        return grid;
    }

    public static lowestPrice = (boardCode: number):number => {
        const pieceTypes: PieceType[] = [PieceType.Seed, PieceType.SmallTree, PieceType.MediumTree, PieceType.LargeTree];
        let lowestPrice = Number.POSITIVE_INFINITY;
        pieceTypes.forEach((pt)=>{
        let currentPrice = PlayerBoard.currentPrice(boardCode, pt);
        lowestPrice = Math.min(lowestPrice,currentPrice);}
        )
        return lowestPrice;
    }

    private static DecodePieceCounts = (boardCode: number, available: number[], onPlayerBoard: number[], max:number) => ({
        available: (boardCode >> available[0]) & available[1],
        onPlayerBoard: (boardCode >> onPlayerBoard[0]) & onPlayerBoard[1],
        max,
        increaseAvailable: () => {
            let currentlyAvailable = (boardCode >> available[0]) & available[1];
            let result = boardCode & (~(available[1] << available[0]));
            result |= (currentlyAvailable + 1) << available[0];
            return result;
        },
        decreaseAvailable: () => {
            let result = boardCode & (~(available[1] << available[0]));
            let currentlyAvailable = (boardCode >> available[0]) & available[1];
            result |= (currentlyAvailable - 1) << available[0];
            return result;
        },
        increaseOnPlayerBoard: () => {
            let currentlyOnPlayerBoard = (boardCode >> onPlayerBoard[0]) & 7;
            let result = boardCode & (~(onPlayerBoard[1] << onPlayerBoard[0]));
            result |= (Math.min(currentlyOnPlayerBoard + 1, max)) << onPlayerBoard[0];
            return result;
        },
        decreaseOnPlayerBoard: () => {
            let result = boardCode & (~(onPlayerBoard[1] << onPlayerBoard[0]));
            let currentlyOnPlayerBoard = (boardCode >> onPlayerBoard[0]) & 7;
            result |= (currentlyOnPlayerBoard - 1) << onPlayerBoard[0];
            return result;
        }
    })

    public static available(boardCode: number):number[] {
        return [PlayerBoard.seeds(boardCode).available,
            PlayerBoard.smallTrees(boardCode).available,
             PlayerBoard.mediumTrees(boardCode).available,
             PlayerBoard.largeTrees(boardCode).available,
        ]
    }

    static currentPrice(boardCode: number, pieceType: PieceType) {
        const onPlayerBoard = PlayerBoard.GetPieces(boardCode, pieceType).onPlayerBoard;
        return onPlayerBoard == 0 ? 0 : PlayerBoard.prices(pieceType)[onPlayerBoard-1];
    }

}

export type PieceDetails = { status: "Ready" | "Filled" | "Empty", price: string, key:string }
